tests/test_types.py

```diff
@@ -0,0 +1,60 @@
+from dffml.df.base import op
+from dffml.df.types import DataFlow, Input, Definition
+from dffml.operation.output import GetSingle
+from dffml.util.asynctestcase import AsyncTestCase
+from dffml.df.memory import MemoryOrchestrator
+from dffml.operation.mapping import MAPPING
+
+Pie = Definition(name="pie", primitive="float", validate=lambda _: 3.14)
+Radius = Definition(name="radius", primitive="float")
+Area = Definition(name="area", primitive="float")
+ShapeName = Definition(
+    name="shape_name", primitive="str", validate=lambda x: x.upper()
+)
+
+
+@op(
+    inputs={"name": ShapeName, "radius": Radius, "pie": Pie},
+    outputs={"shape": MAPPING},
+)
+async def get_circle(name: str, radius: float, pie: float):
+    return {
+        "shape": {
+            "name": name,
+            "radius": radius,
+            "area": pie * radius * radius,
+        }
+    }
+
+
+class TestDefintion(AsyncTestCase):
+    async def test_validate(self):
```

Comment:
Let's add another `test_` where the `pie` datatype raises a `InputValidationError` (add this new error class to `dffml/exceptions.py`).

Use `with self.assertRaises(InputValidationError)` in the new test to make sure that we catch it.